In [1]:
import numpy as np
import pandas as pd
from scipy.sparse import coo_matrix
from sklearn.metrics.pairwise import euclidean_distances
import matplotlib.pyplot as plt

### Dataset: Goodread 10k
Because Collaborative Filtering required a huge dataset of user-item interaction, so we going to use a dataset called "Goodread 10k" which contain 10000 unique books, 53000 unique users and 1 million interaction logs between user-item

https://www.kaggle.com/jealousleopard/goodreadsbooks

In [2]:
rating = pd.read_csv("./data/goodread/ratings.csv")
books = pd.read_csv("./data/goodread/books.csv")
books = books[['id','authors','title','average_rating','ratings_count']].rename(columns={'id':'book_id'})

In [3]:
books.sample(20)

book_id                              authors  \
6716     6717                         Rachel Caine   
8028     8029               Robert Jackson Bennett   
8136     8137                        Tarryn Fisher   
3983     3984                       Gena Showalter   
4754     4755                         Mark Helprin   
4302     4303                          Erin Hunter   
1670     1671                       Rainbow Rowell   
5086     5087                          Lily Koppel   
9172     9173                      Indu Sundaresan   
6886     6887  Sergei Lukyanenko, Andrew Bromfield   
3803     3804                          Dean Koontz   
7155     7156         Lisa Kleypas, Rosalyn Landor   
5684     5685      Orson Scott Card, James Cameron   
1110     1111                           Jane Green   
2403     2404                       Rachel Hartman   
1823     1824                         Sara Shepard   
5403     5404                       Sinclair Lewis   
2068     2069                            Meg Cabot   
9898     9899                     Angela Duckworth   
9474     9475                           Ted Dekker   

                                                  title  average_rating  \
6716  The Morganville Vampires, Volume 3 (The Morgan...            4.39   
8028             City of Stairs (The Divine Cities, #1)            4.13   
8136                                          F*ck Love            4.13   
3983   The Darkest Whisper (Lords of the Underworld #4)            4.33   
4754                                      Winter's Tale            3.50   
4302                   Forest of Secrets (Warriors, #3)            4.40   
1670                                           Carry On            4.19   
5086                           The Astronaut Wives Club            3.26   
9172         The Twentieth Wife (Taj Mahal Trilogy, #1)            4.03   
6886                               Day Watch (Watch #2)            4.05   
3803                                           Whispers            3.90   
7155           Tempt Me at Twilight (The Hathaways, #3)            4.18   
5684                                          The Abyss            4.06   
1110                                          Mr. Maybe            3.41   
2403                          Seraphina (Seraphina, #1)            3.98   
1823                 Flawless (Pretty Little Liars, #2)            3.96   
5403                                        Main Street            3.75   
2068              Queen of Babble (Queen of Babble, #1)            3.71   
9898  Grit: Passion, Perseverance, and the Science o...            4.08   
9474                                               Adam            3.93   

      ratings_count  
6716          13849  
8028          11863  
8136          12203  
3983          35144  
4754          19932  
4302          28913  
1670          46197  
5086          17837  
9172           9873  
6886          13812  
3803          26304  
7155          23125  
5684          16318  
1110          78132  
2403          33510  
1823          59403  
5403          19040  
2068          54248  
9898           7971  
9474           9903

In [4]:
rating.sample(20)

book_id  user_id  rating
84966       850    31376       3
898263     9101    28302       3
278574     2788    40737       4
702974     7073    36656       5
620927     6236    25847       4
888040     8992    28235       2
375763     3763    11569       4
488213     4894      977       4
844223     8533    38889       4
710333     7148    15563       3
883451     8944     1595       2
792275     7991    52095       5
384234     3848    21416       5
813656     8215    34584       2
960720     9772     7254       3
557620     5596    38518       3
322135     3225    17567       3
880179     8908    40241       4
817059     8251    16649       4
55429       555    22496       4

In [5]:
num_users = len(rating['user_id'].unique())
num_books = len(rating['book_id'].unique())
print("Number of user:",num_users)
print("Number of books:",num_books)

Number of user: 53424
Number of books: 10000


![title](./assets/utility-matrix.png)

In [6]:
cor = rating[['user_id','book_id']].values
# -1 because our dataset count index from 1
# but python index count from 0
cor -= 1
val = rating['rating'].values

In [7]:
rating_matrix = coo_matrix((val, (cor[:,0],cor[:,1]))).toarray()

In [8]:
rating_matrix.shape

(53424, 10000)

## Build a Recommendation base on similar users
![title](./assets/alsoview.png)

In [9]:
def get_book_read_by_user(user_id,with_book_info=False):
    user_books = rating[rating['user_id']==user_id].reset_index(drop=True)
    read_books_id = user_books['book_id'].values
    if not with_book_info:
        return read_books_id
    read_books_info = books[books['book_id'].isin(read_books_id)].reset_index(drop=True)
    read_books_info = pd.merge(user_books,read_books_info,on='book_id')
    return read_books_info

In [10]:
def get_similar_user(user_id,k=10):
    # -1 rating table
    _id = user_id - 1
    idx = np.where(rating_matrix[_id]!=0)[0]
    sliced_rating_matrix = rating_matrix[:,idx]
    target = sliced_rating_matrix[_id]
    
    dist = [np.squeeze(euclidean_distances(target.reshape(1,-1),u.reshape(1,-1))) for u in sliced_rating_matrix]
    
    similar_user = np.argsort(dist)
    similar_user += 1
    return similar_user[:k]

In [11]:
def similar_books(user_1,user_2):
    u1 = get_book_read_by_user(user_1)
    u2 = get_book_read_by_user(user_2)
    return np.intersect1d(u1,u2)

In [12]:
target_user = 412
get_book_read_by_user(412,True)

book_id  user_id  rating  \
0      3930      412       4   
1      4640      412       5   
2      4908      412       2   
3      5122      412       4   
4      7379      412       3   
5      7771      412       4   
6      8676      412       5   
7      8703      412       4   
8      8878      412       3   
9      9533      412       4   
10     9759      412       5   

                                              authors  \
0   Aristotle, J.A.K. Thomson, Jonathan Barnes, Hu...   
1                                      Raymond Carver   
2                                        Neil Strauss   
3                                  Hunter S. Thompson   
4                                        Steve Martin   
5                                        Lorrie Moore   
6                                       Richard Price   
7                               Adrian Nicole LeBlanc   
8              Gabriel García Márquez, J.S. Bernstein   
9                                      Sherman Alexie   
10                                         Nick Flynn   

                                                title  average_rating  \
0                              The Nicomachean Ethics            3.91   
1                                           Cathedral            4.30   
2   The Game: Penetrating the Secret Society of Pi...            3.73   
3         Fear and Loathing on the Campaign Trail '72            4.10   
4                          The Pleasure of My Company            3.78   
5                                    Birds of America            4.12   
6                                           Lush Life            3.69   
7   Random Family: Love, Drugs, Trouble, and Comin...            4.23   
8      No One Writes to the Colonel and Other Stories            3.86   
9                                   Reservation Blues            3.98   
10                Another Bullshit Night in Suck City            3.77   

    ratings_count  
0           19380  
1           18378  
2           16698  
3           15306  
4           12378  
5           10768  
6            9478  
7            9053  
8           10480  
9            9919  
10           8441

In [13]:
get_similar_user(412,k=10)

array([  412, 39032,  3222,  2645,   491,  1861,  8284, 28563, 44496,
       24040])

In [16]:
similar_books(412,2645)

array([5122, 7379, 8676])

In [17]:
get_book_read_by_user(39032,True)

book_id  user_id  rating                      authors  \
0         34    39032       2                   E.L. James   
1        106    39032       4                     Tina Fey   
2        144    39032       4            Laura Hillenbrand   
3        190    39032       5               Cheryl Strayed   
4        196    39032       5              Chuck Palahniuk   
..       ...      ...     ...                          ...   
105     8818    39032       5               Raymond Carver   
106     8947    39032       5           John L. Parker Jr.   
107     9038    39032       4                  Diablo Cody   
108     9143    39032       4  Jenna Jameson, Neil Strauss   
109     9759    39032       5                   Nick Flynn   

                                                 title  average_rating  \
0              Fifty Shades of Grey (Fifty Shades, #1)            3.67   
1                                           Bossypants            3.94   
2    Unbroken: A World War II Story of Survival, Re...            4.40   
3    Wild: From Lost to Found on the Pacific Crest ...            3.96   
4                                           Fight Club            4.20   
..                                                 ...             ...   
105                  Will You Please Be Quiet, Please?            4.29   
106                                      Once a Runner            4.05   
107  Candy Girl: A Year in the Life of an Unlikely ...            3.54   
108  How to Make Love Like a Porn Star: A Cautionar...            3.75   
109                Another Bullshit Night in Suck City            3.77   

     ratings_count  
0          1338493  
1           506250  
2           487775  
3           379872  
4           365349  
..             ...  
105           8568  
106           8106  
107          10669  
108           9531  
109           8441  

[110 rows x 7 columns]

## Matrix Factorization

![title](./assets/matrix-factorization.png)


### Why Matrix Factorization

#### 1. Smaller Size

   For example, if we have 53424 users, 10000 items (books) => our rating matrix will have shape of $(U,I) = (53424,10000)$. If each cell is a 'float64' number:

  $$total\_memory = 53424\cdot10000\cdot8$$
  $$total\_memory = 4273920000(b)$$
  $$total\_memory \approx 4,274(gb)$$
  
   If we can factorize this $(U,I)$ matrix to $(U,k)\cdot(I,k)^T$ with $k = 32$ (for example), then we will greatly reduce our space complexity:
   
   $$total\_memory = 53424\cdot32\cdot8 + 10000\cdot32\cdot8$$
   $$total\_memory = 16236544(b)$$
   $$total\_memory \approx 16(mb)$$


#### 2. Predict Items Rating (Matrix completion)
   Make inference time much faster. Instead of calculate similar users to a given target user, then sample a list of recommendation from these similar users, the product of 2 latent matrices will give us predicted rating score of each user to each item.
   
   
#### 3. "Meaningful" latent vectors
   
   From an application point of view, matrix factorization can be used to discover latent features underlying the interactions between two different kinds of entities. (Of course, you can consider more than two kinds of entities and you will be dealing with tensor factorization, which would be more complicated.) And one obvious application is to predict ratings in collaborative filtering
   
   More info: http://nicolas-hug.com/blog/matrix_facto_1

## Singular Value Decomposition (SVD)

![title](./assets/svd.png)

https://en.wikipedia.org/wiki/Singular_value_decomposition

https://research.fb.com/blog/2014/09/fast-randomized-svd/

Note: Definition of SVD in Recommendation System often doesn't equal its definition in Mathematic.  

$A \approx U\cdot\Sigma\cdot{V^T}$

Where:

   - A is our utility matrix that has shape $(m,n)$

   - U is an matrix that has shape $(m,k)$

   - V is an matrix that has shape $(n,k)$
   
   - $\Sigma$ have shape $(k,k)$

In [18]:
from sklearn.utils.extmath import randomized_svd

In [19]:
U, Sigma, V_T = randomized_svd(rating_matrix,n_components=128)
Sigma = np.diag(Sigma)

In [20]:
print(U.shape)
print(Sigma.shape)
print(V_T.shape)

(53424, 128)
(128, 128)
(128, 10000)


In [31]:
# 23122, 42135 (manga)
# 2123 (non-fiction)
# 5231 (author-bias)
# 42152 (cold start)
target_user = 5231
target_user_latent = U[target_user-1].reshape(1,-1)
target_user_latent.shape

(1, 128)

In [32]:
predicted_rating = np.dot(np.dot(target_user_latent,Sigma),V_T)[0]

In [33]:
recommended_books = np.argsort(predicted_rating)[::-1]
recommended_books_rating = np.sort(predicted_rating)[::-1]

recommended_books = recommended_books[:20]
recommended_books_rating = recommended_books_rating[:20]

recommended_books+=1

In [34]:
get_book_read_by_user(target_user,True)

book_id  user_id  rating                            authors  \
0      2771     5231       5                Catherine Ryan Hyde   
1      3012     5231       5                        Donna Mabry   
2      3429     5231       5                   Jennifer McMahon   
3      5135     5231       3                 Mary Higgins Clark   
4      5379     5231       3                      Rachel Abbott   
5      5447     5231       4                      Camron Wright   
6      5628     5231       2          María Dueñas, Daniel Hahn   
7      5738     5231       4                       Chris Cleave   
8      5758     5231       3                       Harlan Coben   
9      5801     5231       3                 Mary Higgins Clark   
10     5807     5231       4                        Darcie Chan   
11     6000     5231       4                      Julie Klassen   
12     6231     5231       5       Bill O'Reilly, Martin Dugard   
13     6742     5231       4                 Mary Higgins Clark   
14     7108     5231       5           Cassie Dandridge Selleck   
15     7239     5231       4                   Katharine Graham   
16     7352     5231       4               James Alexander Thom   
17     7390     5231       3                     Laurie Fabiano   
18     7520     5231       2                  Michelle Richmond   
19     7641     5231       2                          Ivan Doig   
20     7739     5231       4                         Glenn Beck   
21     8030     5231       3                 Mary Higgins Clark   
22     8091     5231       5                      Sandra Dallas   
23     8111     5231       4                      Cynthia Voigt   
24     8324     5231       3  Benedict Freedman, Nancy Freedman   
25     8385     5231       3                    Liza Klaussmann   
26     8475     5231       3                      Sandra Dallas   
27     8514     5231       4                 Mary Higgins Clark   
28     8830     5231       5                 Mary Higgins Clark   
29     9000     5231       2                 Mary Higgins Clark   
30     9096     5231       4                     Charles Martin   
31     9322     5231       1                       Andy Andrews   
32     9348     5231       4                   Jennifer McMahon   
33     9422     5231       5                   Jennifer McMahon   
34     9594     5231       4                   Jennifer McMahon   
35     9760     5231       4                     Beverly  Lewis   

                                                title  average_rating  \
0                                    When I Found You            3.92   
1                                               Maude            4.05   
2                                   The Winter People            3.76   
3                         Loves Music, Loves to Dance            3.89   
4                                         Sleep Tight            3.98   
5                                  The Rent Collector            4.23   
6                                 The Time in Between            4.04   
7                          Everyone Brave is Forgiven            3.80   
8                       Promise Me (Myron Bolitar #8)            3.96   
9                          While My Pretty One Sleeps            3.88   
10                             The Mill River Recluse            3.55   
11                          The Apothecary's Daughter            3.79   
12  Killing Patton: The Strange Death of World War...            4.08   
13                       On the Street Where You Live            3.85   
14                                      The Pecan Man            4.07   
15                                   Personal History            4.13   
16                                   Follow the River            4.21   
17                                   Elizabeth Street            3.90   
18                                    The Year of Fog            3.46   
19                               The Whistling Season           

In [35]:
temp = books[books['book_id'].isin(recommended_books)]
temp.insert(5,'user_predicted_rating',recommended_books_rating)
temp

book_id             authors  \
3017     3018  Mary Higgins Clark   
5013     5014  Mary Higgins Clark   
5076     5077  Mary Higgins Clark   
5134     5135  Mary Higgins Clark   
5796     5797  Mary Higgins Clark   
5800     5801  Mary Higgins Clark   
6145     6146  Mary Higgins Clark   
6303     6304  Mary Higgins Clark   
6714     6715  Mary Higgins Clark   
6741     6742  Mary Higgins Clark   
7569     7570  Mary Higgins Clark   
8001     8002  Mary Higgins Clark   
8011     8012  Mary Higgins Clark   
8029     8030  Mary Higgins Clark   
8513     8514  Mary Higgins Clark   
8829     8830  Mary Higgins Clark   
8999     9000  Mary Higgins Clark   
9067     9068  Mary Higgins Clark   
9240     9241      Gerald N. Lund   
9884     9885  Mary Higgins Clark   

                                                title  average_rating  \
3017                          Where Are the Children?            4.00   
5013                         Two Little Girls in Blue            3.85   
5076                              All Around the Town            3.94   
5134                      Loves Music, Loves to Dance            3.89   
5796                                 You Belong To Me            3.85   
5800                       While My Pretty One Sleeps            3.88   
6145                       Let Me Call You Sweetheart            3.84   
6303                                      Remember Me            3.91   
6714                           A Stranger Is Watching            3.91   
6741                     On the Street Where You Live            3.85   
7569                              Daddy's Little Girl            3.85   
8001                               Where Are You Now?            3.77   
8011                         I Heard That Song Before            3.81   
8029                             The Cradle Will Fall            3.89   
8513                               I'll Be Seeing You            3.79   
8829                               A Cry In The Night            3.84   
8999                        Pretend You Don't See Her            3.86   
9067                            Moonlight Becomes You            3.80   
9240  Truth Will Prevail (The Work and the Glory, #3)            4.30   
9884                                  I'll Walk Alone            3.78   

      ratings_count  user_predicted_rating  
3017          34816               1.542324  
5013          20727               1.507866  
5076          20029               1.481117  
5134          21880               1.474872  
5796          19131               1.471890  
5800          18846               1.463220  
6145          17472               1.442045  
6303          16963               1.413903  
6714          15659               1.410484  
6741          16327               1.399965  
7569          14602               1.376400  
8001          12550               1.305957  
8011          12611               1.299521  
8029          14370               1.277415  
8513          13804               1.148185  
8829          12157               1.033066  
8999          12508               0.946288  
9067          12657               0.726030  
9240           9081               0.566976  
9884           9794               0.282384

# Conclusion

### Collaborative Filtering / Matrix Factorization

#### Pros:

- Smaller memory footprint
   
- Efficient query time (after the factorization step is complete)
   
- "Meaningful" latent space, helpful when compute similar users, similar items
   
- Doesn't require any meta data (item's title, description, tags,... ). Derived users' and items' features directly from the interaction matrix alone.

   
#### Cons:
    
- Popularity bias: Super popular items have more iteraction in the utility matrix => our model will start recommend some super popular items to everyone just because most people have interact with these items.

- Cold-start problems: As oppose to popular items, items that have small number of interactions will never get recommened to anyone. The same thing is true to user, if an user that have small number of interactions, the recommended items will not accurate.
